--- rpython/jit/metainterp/logger.py	Mon Aug 15 15:36:56 2016 +0100
+++ rpython/jit/metainterp/logger.py	Mon Aug 15 18:30:15 2016 +0200
@@ -98,6 +98,7 @@
     def log_abort_loop(self, trace, memo=None):
         debug_start("jit-abort-log")
         if not have_debug_prints():
+            debug_stop("jit-abort-log")
             return
         inputargs, operations = self._unpack_trace(trace)
         logops = self._log_operations(inputargs, operations, ops_offset=None,
